File: programs/apex_racing/src/xp.rs
//! XP math — the authoritative implementation.
//!
//! `src/game/scoring/xp.ts` mirrors this function for function so the number the
//! player sees on the results screen is the number this program writes. Both
//! sides use integer arithmetic in the same order of operations; if you change
//! one, change the other in the same commit.

/// Per-level tuning. Mirrors `src/game/config/levels.ts`.
#[derive(Clone, Copy)]
pub struct LevelParams {
    pub laps: u8,
    /// Target time, ms. The pace term is scored against this.
    pub par_ms: u32,
    /// Fastest physically plausible time. Anything below is rejected outright.
    pub floor_ms: u32,
    pub base_xp: u32,
    /// Drift multiplier in basis points (400 = 0.04x).
    pub drift_bps: u32,
    pub clean_bonus: u32,
    /// Act III gate. 0 means no gate.
    pub drift_target: u32,
    /// Finishing position required to clear the level.
    pub target_position: u8,
    /// XP required to unlock. Enforced at `open_session`.
    pub unlock_xp: u64,
}

/// Indexed by `LEVEL_INDEX` in `src/game/config/levels.ts`. Order is a wire
/// format — append only, never reorder.
pub const LEVELS: [LevelParams; 6] = [
    // All acts use the supplied Suzuka route. Lap counts and timing bounds are
    // calibrated to keep a browser race near the original session length.
    // 0 — ACT I, Cold Start
    LevelParams { laps: 1, par_ms: 249500, floor_ms: 154500, base_xp: 400, drift_bps: 400, clean_bonus: 150, drift_target: 0, target_position: 3, unlock_xp: 0 },
    // 1 — ACT II, Delegation
    LevelParams { laps: 1, par_ms: 243000, floor_ms: 150500, base_xp: 700, drift_bps: 600, clean_bonus: 250, drift_target: 0, target_position: 3, unlock_xp: 0 },
    // 2 — ACT III, Tick Rate
    LevelParams { laps: 1, par_ms: 225000, floor_ms: 139500, base_xp: 850, drift_bps: 1_400, clean_bonus: 300, drift_target: 220, target_position: 3, unlock_xp: 2_500 },
    // 3 — ACT IV, The Commit Window
    LevelParams { laps: 2, par_ms: 477000, floor_ms: 295500, base_xp: 1_100, drift_bps: 900, clean_bonus: 400, drift_target: 0, target_position: 2, unlock_xp: 10_000 },
    // 4 — ACT V, Undelegate
    LevelParams { laps: 1, par_ms: 242500, floor_ms: 150500, base_xp: 1_600, drift_bps: 1_000, clean_bonus: 600, drift_target: 0, target_position: 1, unlock_xp: 30_000 },
    // 5 — Endless Time Attack
    LevelParams { laps: 1, par_ms: 255500, floor_ms: 158500, base_xp: 600, drift_bps: 1_100, clean_bonus: 350, drift_target: 0, target_position: 1, unlock_xp: 0 },
];

/// XP required to unlock each car index. Mirrors `CARS[*].unlockXp`.
pub const CAR_UNLOCK_XP: [u64; 1] = [0];

pub const LEVEL_COUNT: u8 = 6;
pub const CAR_COUNT: u8 = 1;

const PACE_MIN_PCT: u64 = 25;
const PACE_MAX_PCT: u64 = 200;
const PLACING_BONUS: [u32; 6] = [500, 250, 100, 0, 0, 0];
const RISK_PER_DEFERRED_LAP: u64 = 25;
const MAX_RISK_PCT: u64 = 300;

/// Rank thresholds. Mirrors `RANKS` in `src/game/config/progression.ts`.
pub const RANK_THRESHOLDS: [u64; 5] = [0, 2_500, 10_000, 30_000, 75_000];

pub fn rank_for_xp(xp: u64) -> u8 {
    let mut rank = 0u8;
    let mut i = 0usize;
    while i < RANK_THRESHOLDS.len() {
        if xp >= RANK_THRESHOLDS[i] {
            rank = i as u8;
        }
        i += 1;
    }
    rank
}

pub fn level_params(level_id: u8) -> Option<LevelParams> {
    if (level_id as usize) < LEVELS.len() {
        Some(LEVELS[level_id as usize])
    } else {
        None
    }
}

pub fn pace_xp(base_xp: u32, par_ms: u32, total_ms: u32) -> u64 {
    if total_ms == 0 {
        return 0;
    }
    // Multiply before dividing so this is reproducible without floats.
    let ratio_pct = (par_ms as u64 * 100) / total_ms as u64;
    let clamped = ratio_pct.clamp(PACE_MIN_PCT, PACE_MAX_PCT);
    (base_xp as u64 * clamped) / 100
}

pub fn risk_percent(bank_deferred_laps: u8) -> u64 {
    (100 + RISK_PER_DEFERRED_LAP * bank_deferred_laps as u64).min(MAX_RISK_PCT)
}

pub struct XpInput {
    pub total_ms: u32,
    pub drift_score: u32,
    pub collisions: u16,
    pub overtakes: u16,
    pub position: u8,
    pub bank_deferred_laps: u8,
}

pub struct XpBreakdown {
    pub pace: u64,
    pub drift: u64,
    pub clean: u64,
    pub overtakes: u64,
    pub placing: u64,
    pub risk_percent: u64,
    pub total: u64,
}

pub fn compute_xp(level: &LevelParams, input: &XpInput) -> XpBreakdown {
    let pace = pace_xp(level.base_xp, level.par_ms, input.total_ms);
    let drift = (input.drift_score as u64 * level.drift_bps as u64) / 10_000;
    let clean = if input.collisions == 0 {
        level.clean_bonus as u64
    } else {
        0
    };
    let overtakes = input.overtakes as u64 * 25;
    let placing_index = (input.position.clamp(1, 6) - 1) as usize;
    let placing = PLACING_BONUS[placing_index] as u64;

    let subtotal = pace + drift + clean + overtakes + placing;
    let risk = risk_percent(input.bank_deferred_laps);
    let total = (subtotal * risk) / 100;

    XpBreakdown {
        pace,
        drift,
        clean,
        overtakes,
        placing,
        risk_percent: risk,
        total,
    }
}

/// Bounds that make a submitted result *possible*.
///
/// This is not full verification — the client still simulates the race. It
/// closes off the cheap attacks (a one-second lap, a million-point drift score,
/// skipped checkpoints) so forging a result needs a real replay rather than an
/// edited number. Full server-side replay is the next step.
pub fn plausible(
    level: &LevelParams,
    input: &XpInput,
    checkpoints_hit: u16,
    checkpoints_per_lap: u16,
) -> bool {
    if input.total_ms < level.floor_ms {
        return false;
    }
    if input.total_ms as u64 > level.par_ms as u64 * 6 {
        return false;
    }
    if input.position < 1 || input.position > 8 {
        return false;
    }
    // Drift score can only accumulate so fast per millisecond of race time.
    if input.drift_score as u64 > input.total_ms as u64 * 3 / 2 {
        return false;
    }
    let expected = level.laps as u16 * checkpoints_per_lap;
    if checkpoints_per_lap > 0 && checkpoints_hit < expected {
        return false;
    }
    if input.bank_deferred_laps > level.laps {
        return false;
    }
    true
}

File: src/game/scoring/xp.ts
/**
 * XP math.
 *
 * This file is the client half of a pair: `programs/apex_racing/src/xp.rs`
 * implements the identical formula in Rust and is the authority. Both use
 * integer arithmetic in the same order of operations so the number the player
 * sees on the results screen is exactly the number the program writes.
 *
 * If you change anything here, change it there in the same commit.
 */

import type { LevelDefinition } from "../config/levels";
import type { XpBreakdown } from "../types";

/** Bounds on the pace term, in percent. Mirrors the on-chain clamp. */
const PACE_MIN_PCT = 25;
const PACE_MAX_PCT = 200;

/** Placing bonuses for positions 1..6. */
const PLACING_BONUS = [500, 250, 100, 0, 0, 0] as const;

/** Per-lap risk added while a run stays uncommitted (Act IV), in percent. */
export const RISK_PER_DEFERRED_LAP = 25;

/** Cap so deferring cannot run away with the economy. */
const MAX_RISK_PCT = 300;

/**
 * XP forfeited per registered contact, with a barrier or another car.
 *
 * Contact already cost the binary `cleanBonus`; this makes each individual hit
 * cost something, which is what stops a driver from writing off the bonus on lap
 * one and then using the barriers as a guide rail for the rest of the race.
 *
 * Deliberately derived from the collision *count* and nothing else. That count is
 * already streamed to the rollup every tick as `collision_delta` and stored on the
 * session, so the program can compute this term from state it already has — no new
 * instruction argument, no new account field, and no change to a wire format that
 * would need a migration.
 */
export const XP_PER_CONTACT = 120;

/** Penalty in XP for a number of contacts, before it is capped at the subtotal. */
export function penaltyXp(collisions: number): number {
  return Math.max(0, Math.floor(collisions)) * XP_PER_CONTACT;
}

export interface XpInput {
  /** Finish time, ms. */
  totalMs: number;
  driftScore: number;
  collisions: number;
  overtakes: number;
  /** 1-based finishing position. */
  position: number;
  /** Laps completed without banking. */
  bankDeferredLaps: number;
}

/**
 * Pace term: how close the run was to par, clamped.
 *
 * Integer-only, and the multiply happens before the divide so the Rust side can
 * reproduce it exactly without floats.
 */
export function paceXp(baseXp: number, parMs: number, totalMs: number): number {
  if (totalMs <= 0) return 0;
  const ratioPct = Math.floor((parMs * 100) / totalMs);
  const clamped = Math.min(Math.max(ratioPct, PACE_MIN_PCT), PACE_MAX_PCT);
  return Math.floor((baseXp * clamped) / 100);
}

export function riskPercent(bankDeferredLaps: number): number {
  const pct = 100 + RISK_PER_DEFERRED_LAP * Math.max(0, bankDeferredLaps);
  return Math.min(pct, MAX_RISK_PCT);
}

export function computeXp(level: LevelDefinition, input: XpInput): XpBreakdown {
  const pace = paceXp(level.baseXp, level.parMs, input.totalMs);

  // driftMultiplier is authored as a float for tuning readability; convert to
  // integer basis points once so the arithmetic below stays integral.
  const driftBps = Math.round(level.driftMultiplier * 10_000);
  const drift = Math.floor((input.driftScore * driftBps) / 10_000);

  const clean = input.collisions === 0 ? level.cleanBonus : 0;
  const overtakes = Math.max(0, input.overtakes) * 25;
  const placing = PLACING_BONUS[Math.min(Math.max(input.position, 1), 6) - 1] ?? 0;

  const subtotal = pace + drift + clean + overtakes + placing;
  // Capped at the subtotal: the program's arithmetic is unsigned, so a run that
  // earned less than it forfeited has to floor at zero rather than wrap.
  const penalty = Math.min(subtotal, penaltyXp(input.collisions));
  const risk = riskPercent(input.bankDeferredLaps);
  const total = Math.floor(((subtotal - penalty) * risk) / 100);

  return { pace, drift, clean, overtakes, placing, penalty, riskPercent: risk, total };
}

/**
 * Live XP estimate mid-race.
 *
 * Extrapolates the finish time from progress so the HUD number moves smoothly
 * instead of jumping at the line. Deliberately not the authoritative path.
 */
export function projectXp(
  level: LevelDefinition,
  elapsedMs: number,
  lapsCompleted: number,
  lapProgress: number,
  driftScore: number,
  collisions: number,
  overtakes: number,
  position: number,
  bankDeferredLaps: number,
): number {
  const progress = (lapsCompleted + lapProgress) / level.laps;
  if (progress <= 0.01) return 0;
  const projectedTotalMs = Math.max(level.floorMs, Math.floor(elapsedMs / progress));
  return computeXp(level, {
    totalMs: projectedTotalMs,
    driftScore,
    collisions,
    overtakes,
    position,
    bankDeferredLaps,
  }).total;
}

/**
 * Client-side sanity check matching the program's validation.
 *
 * Anything the program would reject should be caught here first so the player
 * gets an explanation instead of a failed transaction.
 */
export function validateResult(
  level: LevelDefinition,
  input: XpInput & { checkpointsHit: number; totalCheckpoints: number },
): { ok: true } | { ok: false; reason: string } {
  if (input.totalMs < level.floorMs) {
    return { ok: false, reason: "Finish time below the physical floor for this track." };
  }
  if (input.totalMs > level.parMs * 6) {
    return { ok: false, reason: "Run exceeded the maximum session duration." };
  }
  const expected = level.laps * input.totalCheckpoints;
  if (input.checkpointsHit < expected) {
    return {
      ok: false,
      reason: `Missed checkpoints: ${input.checkpointsHit}/${expected}.`,
    };
  }
  // A drift score can only grow so fast: cap is score-per-ms times duration.
  const maxDrift = Math.floor(input.totalMs * 1.5);
  if (input.driftScore > maxDrift) {
    return { ok: false, reason: "Drift score exceeds the per-tick ceiling." };
  }
  if (input.position < 1 || input.position > 8) {
    return { ok: false, reason: "Invalid finishing position." };
  }
  return { ok: true };
}
